frontend/src/global/api.ts

```diff
@@ -16,7 +17,10 @@ const config = new Configuration({
         if (context.response.status === 403) {
           // Failed requests return 403 and mean that the access token is
           // expired, so we trigger a page refresh.
-          window.location.reload(false);
+          const authenticated = await checkAuthAndRefreshToken();
+          if (!authenticated) {
+            window.location.reload(false);
+          }
```

Comment:
With this change if you ever get a 403 error, the app will assume it's because the token has expired, and just use the refresh token to get a new one. It should work seamlessly, without refresh. 
You can test this by logging in, and then editing your access token to an invalid value, with the next request, your login token will be refreshed.

Reply:
So, there's a few failing edge cases here:
* Any requests made with data will lose context
* If you go to the console after the access is expired, the interface just keeps on loading forever.

I've done a few tests locally, and manage to add a bit of code that **retries** the request if the token is successfully updated. Here is the entire snippet:
```
const getApiKey = () => {
  const authToken = localStorage.getItem('token_access');
  // The current version of the API client requires us to specify the word "Bearer"
  // The previous version did not.
  return authToken ? `Bearer ${authToken}` : '';
}

const config = new Configuration({
  basePath: `${OCIM_API_BASE}/api`,
  apiKey: (name: string, scopes?: string[]) => {
    return getApiKey();
  },
  middleware: [
    {
      post: async context => {
        if (context.response.status === 403) {
          // Failed requests return 403 and mean that the access token is
          // expired, so we trigger a page refresh.
          const authenticated = await checkAuthAndRefreshToken();
          if (!authenticated) {
            window.location.reload(false);
          }
          else {
            // Retry request after updating API key
            const requestContext = context.init;
            requestContext.headers = {
              Authorization: getApiKey()
            };
            let retry = context.fetch(context.url, requestContext);
            return retry;
          }
        } else if (context.response.status === 401) {
          // If the refresh endpoint fails, it means that the refresh
          // is expired, and the user is effectively logged out.
          performLogout();
        }
        return context.response;
      }
    }
  ]
});
``` 

Can you include this in your diff?

Reply:
I tried the exact same thing but it wasn't working. Now I realise the obvious stupid mistake I made, which is that obviously the old request will have the old token so it needs to be updated!